streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java

```diff
@@ -495,57 +477,69 @@ public String queryableStoreName() {
         } else if (!rightOuter) { // left
-            joinThis = new KTableKTableLeftJoin<>(this, (KTableImpl<K, ?, V1>) other, joiner);
-            joinOther = new KTableKTableRightJoin<>((KTableImpl<K, ?, V1>) other, this, reverseJoiner(joiner));
+            joinThis = new KTableKTableLeftJoin<>(this, (KTableImpl<K, ?, VO>) other, joiner);
+            joinOther = new KTableKTableRightJoin<>((KTableImpl<K, ?, VO>) other, this, reverseJoiner(joiner));
         } else { // outer
-            joinThis = new KTableKTableOuterJoin<>(this, (KTableImpl<K, ?, V1>) other, joiner);
-            joinOther = new KTableKTableOuterJoin<>((KTableImpl<K, ?, V1>) other, this, reverseJoiner(joiner));
+            joinThis = new KTableKTableOuterJoin<>(this, (KTableImpl<K, ?, VO>) other, joiner);
+            joinOther = new KTableKTableOuterJoin<>((KTableImpl<K, ?, VO>) other, this, reverseJoiner(joiner));
         }
 
-        final KTableKTableJoinMerger<K, R> joinMerge = new KTableKTableJoinMerger<>(joinThis, joinOther, internalQueryableName);
+        final String joinThisName = builder.newProcessorName(JOINTHIS_NAME);
+        final String joinOtherName = builder.newProcessorName(JOINOTHER_NAME);
 
-        final KTableKTableJoinNode.KTableKTableJoinNodeBuilder<K, V, V1, R> kTableJoinNodeBuilder = KTableKTableJoinNode.kTableKTableJoinNodeBuilder();
+        final ProcessorParameters<K, Change<V>> joinThisProcessorParameters = new ProcessorParameters<>(joinThis, joinThisName);
+        final ProcessorParameters<K, Change<VO>> joinOtherProcessorParameters = new ProcessorParameters<>(joinOther, joinOtherName);
+
+        final KTableKTableJoinNode.KTableKTableJoinNodeBuilder<K, V, VO, VR> kTableJoinNodeBuilder =
+            KTableKTableJoinNode.<K, V, VO, VR>kTableKTableJoinNodeBuilder(builder)
+                .withNodeName(joinMergeName)
+                .withJoinThisProcessorParameters(joinThisProcessorParameters)
+                .withJoinOtherProcessorParameters(joinOtherProcessorParameters)
+                .withThisJoinSideNodeName(name)
+                .withOtherJoinSideNodeName(((KTableImpl) other).name)
+                .withJoinThisStoreNames(valueGetterSupplier().storeNames())
+                .withJoinOtherStoreNames(((KTableImpl) other).valueGetterSupplier().storeNames());
 
-        // only materialize if specified in Materialized
         if (materializedInternal != null) {
-            kTableJoinNodeBuilder.withMaterializedInternal(materializedInternal);
+            final Serde<K> keySerde = materializedInternal.keySerde() != null ? materializedInternal.keySerde() : this.keySerde;
+            final Serde<VR> valueSerde = materializedInternal.valueSerde();
+            final String queryableStoreName = materializedInternal.storeName();
+            final StoreBuilder<KeyValueStore<K, VR>> storeBuilder = new KeyValueStoreMaterializer<>(materializedInternal).materialize();
+            kTableJoinNodeBuilder
```

Comment:
This part and the line 525-529 below can be extracted out of if condition.

Reply:
I tried to, but could not; all parameter variables to `KTableKTableJoinNodeBuilder` must be `final`, but Java does not support conditional instantiation like Scala. 

One alternative approach would be introducing a new class like `MaterializeConfig` to contain them. What do you think?

Reply:
@dongjinleekr you could do something like:
```java
final Serde<K> keySerde;
final Serde<VR> valueSerde;
final String queryableStoreName;
final StoreBuilder<KeyValueStore<K, VR>> storeBuilder;

if (materializedInternal != null){
   ....
else {
   ....
}

kTableJoinNodeBuilder
       .withKeySerde(keySerde)
       .withValueSerde(valueSerde)
       .withQueryableStoreName(queryableStoreName)
       .withStoreBuilder(storeBuilder);
```
Alternatively you can make them non-`final` and set the default values and update the variables inside the `if` condition, and simply remove the `else` block.  It's a matter of personal style so I'll leave it up to you.